## Data Description

This research employed a binary variable, default payment (Yes = 1, No = 0), as the response variable. This study reviewed the literature and used the following 23 variables as explanatory variables:

    X1 (LIMIT_BAL): Amount of the given credit (NT dollar)
    X2 (SEX): Gender (1 = male; 2 = female).
    X3 (EDUCATION): Education (1 = graduate school; 2 = university; 3 = high school; 4 = others).
    X4 (MARRIAGE): Marital status (1 = married; 2 = single; 3 = others).
    X5 (AGE): in years.
    X6 - X11 (PAY_1-PAY_6): History of past payment. PAY_1 - Sept, PAY_2 - Aug, .. , PAY_6 - Apr
        
    The measurement scale for the repayment status is: -2 = no credit used, -1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above.
        
    X12-X17 (BILL_AMT1-BILL_AMT6): Amount of bill statement (NT dollar).
        
    X18-X23 (PAY_AMT1-PAY_AMT6): Amount of previous payment (NT dollar). 

## Business Understanding

This data is provided by a Taiwanees Credit Card Company. It contains recent financial data for 30,000 account holders. The rows are labeled by whether the month after a six month historical data period, the account holder has defaulted.

## Data Preperation

In [1]:
import pandas as pd
import numpy as np

In [10]:
df = pd.read_excel("default_of_credit_card_clients__updated.xls")
#df.rename(columns=df.iloc[0])

In [11]:
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0  798fc410-45c1      20000    2          2         1   24     2      2   
1  8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2  85698822-43f5      90000    2          2         2   34     0      0   
3  0737c11b-be42      50000    2          2         1   37     0      0   
4  3b7f77cc-dbc0      50000    1          2         1   57    -1      0   

   PAY_3  PAY_4             ...              BILL_AMT4  BILL_AMT5  BILL_AMT6  \
0     -1     -1             ...                      0          0          0   
1      0      0             ...                   3272       3455       3261   
2      0      0             ...                  14331      14948      15549   
3      0      0             ...                  28314      28959      29547   
4     -1      0             ...                  20940      19146      19131   

   PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
0         0       689         0         0         0         0   
1         0      1000      1000      1000         0      2000   
2      1518      1500      1000      1000      1000      5000   
3      2000      2019      1200      1100      1069      1000   
4      2000     36681     10000      9000       689       679   

   default payment next month  
0                           1  
1                           1  
2                           0  
3                           0  
4                           0  

[5 rows x 25 columns]

In [12]:
df.shape

(30000, 25)

In [17]:
df['ID'].nunique()

29687

There seems to be some duplicates in the data. 

In [18]:
id_counts = df['ID'].value_counts()

In [24]:
id_counts.value_counts()

1    29374
2      313
Name: ID, dtype: int64

313 IDs occur twice.

In [25]:
#Boolean Mask
bol_counts = id_counts ==2

In [26]:
bol_counts[:5]

8a6bbf3d-8a2d    True
98b8a2ca-8988    True
527c11ea-113e    True
63b20368-1a04    True
555a25cd-2d05    True
Name: ID, dtype: bool

In [28]:
dup_ids = id_counts.index[bol_counts]

In [31]:
len(dup_ids)

313

In [33]:
df_dup = df[df['ID'].isin(dup_ids)]

In [34]:
df_dup.head(10)

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE          PAY_1  \
141  26bde6da-f148     180000    1          2         2   28             -1   
196  42009c72-651f     150000    2          2         1   34             -2   
241  26bde6da-f148          0    0          0         0    0              0   
296  42009c72-651f          0    0          0         0    0              0   
323  73ed1381-3a3f     230000    2          2         1   28              0   
340  4f95b36b-ab10     240000    2          2         2   30             -2   
398  8d3a2576-a958     210000    2          2         1   30  Not available   
423  73ed1381-3a3f          0    0          0         0    0              0   
440  4f95b36b-ab10          0    0          0         0    0              0   
498  8d3a2576-a958          0    0          0         0    0              0   

     PAY_2  PAY_3  PAY_4             ...              BILL_AMT4  BILL_AMT5  \
141     -1     -1     -1             ...                    332        416   
196     -2     -2     -2             ...                    116          0   
241      0      0      0             ...                      0          0   
296      0      0      0             ...                      0          0   
323      0      0     -2             ...                   2100        767   
340     -2     -2     -2             ...                      0         92   
398      0      0      0             ...                  91143      82733   
423      0      0      0             ...                      0          0   
440      0      0      0             ...                      0          0   
498      0      0      0             ...                      0          0   

     BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
141        416         0       416       332       500      3500       832   
196       1500         0         0       116         0      1500         0   
241          0         0         0         0         0         0         0   
296          0         0         0         0         0         0         0   
323       6173      1710      1100      1000       767      6173      1000   
340          0        92        92         0        92         0      1883   
398      74125      4002      5000      3500      3000      2700      5000   
423          0         0         0         0         0         0         0   
440          0         0         0         0         0         0         0   
498          0         0         0         0         0         0         0   

     default payment next month  
141                           0  
196                           0  
241                           0  
296                           0  
323                           1  
340                           0  
398                           0  
423                           0  
440                           0  
498                           0  

[10 rows x 25 columns]

Seems like the duplicate rows have 0 for LIMIT_BAL etc.

In [35]:
zero_mask = df == 0

In [36]:
len(zero_mask)

30000

In [38]:
zero_mask.head()

ID  LIMIT_BAL    SEX  EDUCATION  MARRIAGE    AGE  PAY_1  PAY_2  PAY_3  \
0  False      False  False      False     False  False  False  False  False   
1  False      False  False      False     False  False  False  False   True   
2  False      False  False      False     False  False   True   True   True   
3  False      False  False      False     False  False   True   True   True   
4  False      False  False      False     False  False  False   True  False   

   PAY_4             ...              BILL_AMT4  BILL_AMT5  BILL_AMT6  \
0  False             ...                   True       True       True   
1   True             ...                  False      False      False   
2   True             ...                  False      False      False   
3   True             ...                  False      False      False   
4   True             ...                  False      False      False   

   PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
0      True     False      True      True      True      True   
1      True     False     False     False      True     False   
2     False     False     False     False     False     False   
3     False     False     False     False     False     False   
4     False     False     False     False     False     False   

   default payment next month  
0                       False  
1                       False  
2                        True  
3                        True  
4                        True  

[5 rows x 25 columns]

In [40]:
#Check among all rows where all columns are zero (except id)

row_zero_mask = zero_mask.iloc[:,1:].all(axis=1)

In [41]:
row_zero_mask.head()

0    False
1    False
2    False
3    False
4    False
dtype: bool

In [42]:
sum(row_zero_mask)

315

So there are 315 rows that have zeros for every column excpet the ID. Since this number is greater than the number of duplicates (313), eliminating the 315 could solve both issues.

In [43]:
clean_df = df.loc[~row_zero_mask]

In [46]:
clean_df.shape

(29685, 25)

In [47]:
clean_df['ID'].nunique()

29685

In [48]:
clean_df.describe()

LIMIT_BAL           SEX     EDUCATION      MARRIAGE           AGE  \
count    29685.000000  29685.000000  29685.000000  29685.000000  29685.000000   
mean    167519.948796      1.603335      1.852788      1.552198     35.481354   
std     129716.478534      0.489214      0.789465      0.521869      9.212199   
min      10000.000000      1.000000      0.000000      0.000000     21.000000   
25%      50000.000000      1.000000      1.000000      1.000000     28.000000   
50%     140000.000000      2.000000      2.000000      2.000000     34.000000   
75%     240000.000000      2.000000      2.000000      2.000000     41.000000   
max    1000000.000000      2.000000      6.000000      3.000000     79.000000   

              PAY_2         PAY_3         PAY_4         PAY_5         PAY_6  \
count  29685.000000  29685.000000  29685.000000  29685.000000  29685.000000   
mean      -0.134277     -0.166077     -0.221627     -0.266700     -0.291090   
std        1.197440      1.197278      1.168278      1.133156      1.150653   
min       -2.000000     -2.000000     -2.000000     -2.000000     -2.000000   
25%       -1.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        8.000000      8.000000      8.000000      8.000000      8.000000   

                  ...                  BILL_AMT4      BILL_AMT5  \
count             ...               29685.000000   29685.000000   
mean              ...               43245.439279   40307.628263   
std               ...               64276.899483   60787.204536   
min               ...             -170000.000000  -81334.000000   
25%               ...                2329.000000    1760.000000   
50%               ...               19040.000000   18106.000000   
75%               ...               54515.000000   50244.000000   
max               ...              891586.000000  927171.000000   

           BILL_AMT6       PAY_AMT1      PAY_AMT2       PAY_AMT3  \
count   29685.000000   29685.000000  2.968500e+04   29685.000000   
mean    38888.682095    5672.886812  5.917545e+03    5229.295570   
std     59588.111810   16616.452239  2.310628e+04   17650.360969   
min   -339603.000000       0.000000  0.000000e+00       0.000000   
25%      1257.000000    1000.000000  8.320000e+02     390.000000   
50%     17067.000000    2100.000000  2.008000e+03    1800.000000   
75%     49252.000000    5007.000000  5.000000e+03    4508.000000   
max    961664.000000  873552.000000  1.684259e+06  896040.000000   

            PAY_AMT4       PAY_AMT5       PAY_AMT6  default payment next month  
count   29685.000000   29685.000000   29685.000000                29685.000000  
mean     4826.770827    4805.203840    5219.023008                    0.221054  
std     15607.256127   15311.798994   17798.365183                    0.414964  
min         0.000000       0.000000       0.000000                    0.000000  
25%       296.000000     250.000000     115.000000                    0.000000  
50%      1500.000000    1500.000000    1500.000000                    0.000000  
75%      4015.000000    4041.000000    4000.000000                    0.000000  
max    621000.000000  426529.000000  528666.000000                    1.000000  

[8 rows x 23 columns]

In [49]:
clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29685 entries, 0 to 29999
Data columns (total 25 columns):
ID                            29685 non-null object
LIMIT_BAL                     29685 non-null int64
SEX                           29685 non-null int64
EDUCATION                     29685 non-null int64
MARRIAGE                      29685 non-null int64
AGE                           29685 non-null int64
PAY_1                         29685 non-null object
PAY_2                         29685 non-null int64
PAY_3                         29685 non-null int64
PAY_4                         29685 non-null int64
PAY_5                         29685 non-null int64
PAY_6                         29685 non-null int64
BILL_AMT1                     29685 non-null int64
BILL_AMT2                     29685 non-null int64
BILL_AMT3                     29685 non-null int64
BILL_AMT4                     29685 non-null int64
BILL_AMT5                     29685 non-null int64
BILL_AMT6             

PAY_1 is of type object instead of int64. Let's have a look!

In [50]:
clean_df['PAY_1'].value_counts()

0                13087
-1                5047
1                 3261
Not available     3021
-2                2476
2                 2378
3                  292
4                   63
5                   23
8                   17
6                   11
7                    9
Name: PAY_1, dtype: int64

Contains 'Not Available'. Let's remove rows with this.

In [51]:
na_mask = clean_df['PAY_1'] != 'Not available'

In [52]:
clean_df2 = clean_df.loc[na_mask]

In [53]:
clean_df2['PAY_1'].value_counts()

 0    13087
-1     5047
 1     3261
-2     2476
 2     2378
 3      292
 4       63
 5       23
 8       17
 6       11
 7        9
Name: PAY_1, dtype: int64

In [54]:
clean_df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26664 entries, 0 to 29999
Data columns (total 25 columns):
ID                            26664 non-null object
LIMIT_BAL                     26664 non-null int64
SEX                           26664 non-null int64
EDUCATION                     26664 non-null int64
MARRIAGE                      26664 non-null int64
AGE                           26664 non-null int64
PAY_1                         26664 non-null object
PAY_2                         26664 non-null int64
PAY_3                         26664 non-null int64
PAY_4                         26664 non-null int64
PAY_5                         26664 non-null int64
PAY_6                         26664 non-null int64
BILL_AMT1                     26664 non-null int64
BILL_AMT2                     26664 non-null int64
BILL_AMT3                     26664 non-null int64
BILL_AMT4                     26664 non-null int64
BILL_AMT5                     26664 non-null int64
BILL_AMT6             

In [55]:
#Convert from object to int64
clean_df2['PAY_1'] = clean_df2['PAY_1'].astype('int64')

C:\Users\tusha\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [56]:
clean_df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26664 entries, 0 to 29999
Data columns (total 25 columns):
ID                            26664 non-null object
LIMIT_BAL                     26664 non-null int64
SEX                           26664 non-null int64
EDUCATION                     26664 non-null int64
MARRIAGE                      26664 non-null int64
AGE                           26664 non-null int64
PAY_1                         26664 non-null int64
PAY_2                         26664 non-null int64
PAY_3                         26664 non-null int64
PAY_4                         26664 non-null int64
PAY_5                         26664 non-null int64
PAY_6                         26664 non-null int64
BILL_AMT1                     26664 non-null int64
BILL_AMT2                     26664 non-null int64
BILL_AMT3                     26664 non-null int64
BILL_AMT4                     26664 non-null int64
BILL_AMT5                     26664 non-null int64
BILL_AMT6              

In [57]:
clean_df2.to_excel("clean_data.xls")